nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs

```diff
@@ -3084,14 +3084,85 @@ impl ArrayPrototype {
+    /// their order within the array is the same as the order in which they appear
+    /// in the argument list.
+    ///
+    /// > ### Note
+    /// >
+    /// > This method is intentionally generic; it does not require that its
+    /// > this value be an Array. Therefore it can be transferred to other
+    /// > kinds of objects for use as a method.
     fn unshift(
-        _agent: &mut Agent,
-        _gc: GcScope<'_, '_>,
-
-        _this_value: Value,
-        _: ArgumentsList,
+        agent: &mut Agent,
+        mut gc: GcScope<'_, '_>,
+        this_value: Value,
+        items: ArgumentsList,
     ) -> JsResult<Value> {
-        todo!();
+        // 1. Let O be ? ToObject(this value).
+        let o = to_object(agent, this_value)?;
+        // 2. Let len be ? LengthOfArrayLike(O).
+        let len = length_of_array_like(agent, gc.reborrow(), o)?;
+        // 3. Let argCount be the number of elements in items.
+        let arg_count = items.len();
+        // 4. If argCount > 0, then
+        if len > 0 {
+            // a. If len + argCount > 2**53 - 1, throw a TypeError exception.
+            if (len + arg_count as i64) > SmallInteger::MAX_NUMBER {
+                return Err(agent.throw_exception_with_static_message(
+                    ExceptionType::TypeError,
+                    "Array length overflow",
+                ));
+            }
+            // b. Let k be len.
+            let mut k = len;
+            // c. Repeat, while k > 0,
+            while k > 0 {
+                // i. Let from be ! ToString(𝔽(k - 1)).
+                let from = (k - 1).try_into().unwrap();
+                // ii. Let to be ! ToString(𝔽(k + argCount - 1)).
+                let to = (k + arg_count as i64 - 1).try_into().unwrap();
+                // iii. Let fromPresent be ? HasProperty(O, from).
+                let from_present = has_property(agent, gc.reborrow(), o, from)?;
+                // iv. If fromPresent is true, then
+                if from_present {
+                    // 1. Let fromValue be ? Get(O, from).
+                    let from_value = get(agent, gc.reborrow(), o, from)?;
+                    // 2. Perform ? Set(O, to, fromValue, true).
+                    set(agent, gc.reborrow(), o, to, from_value, true)?;
+                } else {
+                    // v. Else,
+                    // 1. Assert: fromPresent is false.
+                    // 2. Perform ? DeletePropertyOrThrow(O, to).
+                    delete_property_or_throw(agent, gc.reborrow(), o, to)?;
+                }
+                // vi. Set k to k - 1.
+                k -= 1;
+            }
+            // d. Let j be +0𝔽.
+            let mut j = 0;
+            // e. For each element E of items, do
+            for e in items.iter() {
+                // i. Perform ? Set(O, ! ToString(j), E, true).
+                set(agent, gc.reborrow(), o, j.try_into().unwrap(), *e, true)?;
```

Comment:
clippy is giving up an error here as well, if we cant fix it `#[allow(clippy::unnecessary_fallible_conversions)]` this should work

Reply:
Just `into()`

Reply:
yeah that seems to work on my end